chia-client/src/peer.rs

```diff
@@ -0,0 +1,207 @@
+use std::io::Cursor;
+use std::{collections::HashMap, sync::Arc};
+
+use chia_protocol::{
+    ChiaProtocolMessage, CoinStateUpdate, Handshake, Message, NewPeakWallet, NodeType,
+    ProtocolMessageTypes,
+};
+use chia_traits::Streamable;
+use futures_util::stream::SplitSink;
+use futures_util::{SinkExt, StreamExt};
+use tokio::sync::{broadcast, oneshot, Mutex};
+use tokio::{net::TcpStream, task::JoinHandle};
+use tokio_tungstenite::{MaybeTlsStream, WebSocketStream};
+
+use crate::{Error, Result};
+
+type WebSocket = WebSocketStream<MaybeTlsStream<TcpStream>>;
+type Requests = Arc<Mutex<HashMap<u16, oneshot::Sender<Message>>>>;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PeerEvent {
+    CoinStateUpdate(CoinStateUpdate),
+    NewPeakWallet(NewPeakWallet),
+}
+
+pub struct Peer {
+    sink: Mutex<SplitSink<WebSocket, tungstenite::Message>>,
+    inbound_task: JoinHandle<()>,
+    event_receiver: broadcast::Receiver<PeerEvent>,
+    requests: Requests,
+    nonce: Mutex<u16>,
+}
+
+impl Peer {
+    pub fn new(ws: WebSocket) -> Self {
```

Comment:
if `WebSocket` would be a template parameter, we can test this really well, with some helping infrastructure

Reply:
I'm not sure how to do use a generic Sink + Stream without losing type information that's necessary to receive messages.

Reply:
Perhaps we can revisit this in the future